File: tests/lua/cartridge_controller.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { writeLE32 } from '../../machine/ts/common/endian';
import {
	DMA_CONTROL_BLOCK_WORDS_SHIFT,
	DMA_CONTROL_READ_INCREMENT,
	DMA_CONTROL_READ_REQUEST_SHIFT,
	DMA_CONTROL_WRITE_INCREMENT,
	DMA_CONTROL_WRITE_REQUEST_SHIFT,
	DMA_REQUEST_CARTRIDGE_SLOT0_WRITE,
	DMA_REQUEST_CARTRIDGE_SLOT1_READ,
	DMA_TRIGGER_START,
	IO_CART_SELECT,
	IO_CART_SLOT0_BOARD,
	IO_CART_SLOT0_RAM_BYTES,
	IO_CART_SLOT1_BOARD,
	IO_CART_SLOT1_RAM_BYTES,
	IO_CART_STATUS,
	IO_DMA0_CONTROL,
	IO_DMA0_READ_ADDR,
	IO_DMA0_TRANSFER_COUNT,
	IO_DMA0_TRIGGER,
	IO_DMA0_WRITE_ADDR,
	IO_IRQ_ACK,
	IO_IRQ_FLAGS,
	IRQ_CARTRIDGE_SLOT0,
	IRQ_CARTRIDGE_SLOT1,
} from '../../machine/ts/machine/bus/io';
import { CPU } from '../../machine/ts/machine/cpu/cpu';
import {
	CARTRIDGE_BOARD_MAILBOX,
	CARTRIDGE_BOARD_RAM,
	CARTRIDGE_MAILBOX_CONTROL_DREQ_READ,
	CARTRIDGE_MAILBOX_CONTROL_DREQ_WRITE,
	CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER,
	CARTRIDGE_MAILBOX_CONTROL_OFFSET,
	CARTRIDGE_MAILBOX_DATA_OFFSET,
	CARTRIDGE_MAILBOX_IRQ_ACK_OFFSET,
	CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING,
	CARTRIDGE_MAILBOX_STATUS_OFFSET,
	CARTRIDGE_STATUS_SELECTED_SLOT1,
	CARTRIDGE_STATUS_SLOT0_PRESENT,
	CARTRIDGE_STATUS_SLOT0_PROGRAM,
	CARTRIDGE_STATUS_SLOT1_PRESENT,
	CARTRIDGE_STATUS_SLOT1_PROGRAM,
	type CartridgeSlotMediaPair,
} from '../../machine/ts/machine/devices/cartridge/contracts';
import { DmaController } from '../../machine/ts/machine/devices/dma/controller';
import { IrqController } from '../../machine/ts/machine/devices/irq/controller';
import {
	CART_MMIO_BASE,
	CART_RAM_BASE,
	CART_ROM_BASE,
} from '../../machine/ts/machine/memory/map';
import { Memory } from '../../machine/ts/machine/memory/memory';
import { DeviceScheduler } from '../../machine/ts/machine/scheduler/device';

const MAILBOX_DATA_ADDRESS = CART_MMIO_BASE + CARTRIDGE_MAILBOX_DATA_OFFSET;
const MAILBOX_CONTROL_ADDRESS = CART_MMIO_BASE + CARTRIDGE_MAILBOX_CONTROL_OFFSET;
const MAILBOX_STATUS_ADDRESS = CART_MMIO_BASE + CARTRIDGE_MAILBOX_STATUS_OFFSET;
const MAILBOX_IRQ_ACK_ADDRESS = CART_MMIO_BASE + CARTRIDGE_MAILBOX_IRQ_ACK_OFFSET;

type CartridgeHarness = {
	memory: Memory;
	cpu: CPU;
	dma: DmaController;
	irq: IrqController;
	scheduler: DeviceScheduler;
};

function createHarness(cartridgeSlots: CartridgeSlotMediaPair): CartridgeHarness {
	const memory = new Memory({
		systemRom: new Uint8Array(0),
		cartridgeSlots,
	});
	const irq = new IrqController(memory);
	const cpu = new CPU(memory, irq);
	const scheduler = new DeviceScheduler(cpu);
	const dma = new DmaController(memory, cpu, irq, scheduler);
	memory.cartridgeController.connect(memory, irq, dma);
	irq.reset();
	dma.reset();
	memory.cartridgeController.reset();
	dma.setTiming(1, 0, 1, 1, 0, scheduler.nowCycles);
	return { memory, cpu, dma, irq, scheduler };
}

function slot(
	romWords: readonly number[],
	boardWord: number,
	ramByteCount: number,
	present: boolean,
	programPresent: boolean,
): CartridgeSlotMediaPair[number] {
	const rom = new Uint8Array(romWords.length * 4);
	for (let index = 0; index < romWords.length; index += 1) {
		writeLE32(rom, index * 4, romWords[index]!);
	}
	return { rom, boardWord, ramByteCount, present, programPresent };
}

test('cartridge bus selects one physical socket and retains the raw selection latch', () => {
	const slot0 = slot([0x11223344], CARTRIDGE_BOARD_RAM, 16, true, true);
	const slot1 = slot([0xaabbccdd], CARTRIDGE_BOARD_MAILBOX, 0, true, true);
	const { memory } = createHarness([slot0, slot1]);

	assert.equal(memory.readMappedU32LE(IO_CART_SELECT), 0);
	assert.equal(memory.readMappedU32LE(CART_ROM_BASE), 0x11223344);
	assert.equal(
		memory.readMappedU32LE(IO_CART_STATUS),
		CARTRIDGE_STATUS_SLOT0_PRESENT
			| CARTRIDGE_STATUS_SLOT1_PRESENT
			| CARTRIDGE_STATUS_SLOT0_PROGRAM
			| CARTRIDGE_STATUS_SLOT1_PROGRAM,
	);
	assert.equal(memory.readMappedU32LE(IO_CART_SLOT0_BOARD), CARTRIDGE_BOARD_RAM);
	assert.equal(memory.readMappedU32LE(IO_CART_SLOT0_RAM_BYTES), 16);
	assert.equal(memory.readMappedU32LE(IO_CART_SLOT1_BOARD), CARTRIDGE_BOARD_MAILBOX);
	assert.equal(memory.readMappedU32LE(IO_CART_SLOT1_RAM_BYTES), 0);

	memory.writeMappedU32LE(IO_CART_SELECT, 0xa5a50001);
	assert.equal(memory.readMappedU32LE(IO_CART_SELECT), 0xa5a50001);
	assert.equal(memory.readMappedU32LE(CART_ROM_BASE), 0xaabbccdd);
	assert.equal(
		memory.readMappedU32LE(IO_CART_STATUS),
		CARTRIDGE_STATUS_SLOT0_PRESENT
			| CARTRIDGE_STATUS_SLOT1_PRESENT
			| CARTRIDGE_STATUS_SLOT0_PROGRAM
			| CARTRIDGE_STATUS_SLOT1_PROGRAM
			| CARTRIDGE_STATUS_SELECTED_SLOT1,
	);
});

test('cartridge RAM, mailbox state, reset, and restore remain socket-local', () => {
	const board = CARTRIDGE_BOARD_RAM | CARTRIDGE_BOARD_MAILBOX;
	const { memory } = createHarness([
		slot([], board, 16, true, false),
		slot([], board, 16, true, true),
	]);

	assert.equal(memory.readMappedU32LE(IO_CART_SELECT), 1);
	memory.writeMappedU32LE(CART_RAM_BASE, 0x11112222);
	memory.writeMappedU32LE(MAILBOX_DATA_ADDRESS, 0x33334444);
	memory.writeMappedU32LE(
		MAILBOX_CONTROL_ADDRESS,
		0x80000003,
	);
	assert.equal(
		memory.readMappedU32LE(MAILBOX_CONTROL_ADDRESS),
		0x80000002,
	);
	assert.equal(memory.readMappedU32LE(MAILBOX_STATUS_ADDRESS), CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING);
	assert.notEqual(memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT1, 0);

	memory.writeMappedU32LE(IO_CART_SELECT, 0x2468ace0);
	memory.writeMappedU32LE(CART_RAM_BASE, 0x55556666);
	memory.writeMappedU32LE(MAILBOX_DATA_ADDRESS, 0x77778888);
	memory.writeMappedU32LE(
		MAILBOX_CONTROL_ADDRESS,
		CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER | CARTRIDGE_MAILBOX_CONTROL_DREQ_WRITE,
	);
	assert.notEqual(memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT0, 0);
	const saved = memory.cartridgeController.captureState();

	memory.writeMappedU32LE(CART_RAM_BASE, 0);
	memory.writeMappedU32LE(MAILBOX_DATA_ADDRESS, 0);
	memory.writeMappedU32LE(MAILBOX_IRQ_ACK_ADDRESS, 1);
	memory.writeMappedU32LE(IO_CART_SELECT, 1);
	memory.cartridgeController.restoreState(saved);

	assert.equal(memory.readMappedU32LE(IO_CART_SELECT), 0x2468ace0);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE), 0x55556666);
	assert.equal(memory.readMappedU32LE(MAILBOX_DATA_ADDRESS), 0x77778888);
	assert.equal(memory.readMappedU32LE(MAILBOX_STATUS_ADDRESS), CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING);
	memory.writeMappedU32LE(IO_CART_SELECT, 1);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE), 0x11112222);
	assert.equal(memory.readMappedU32LE(MAILBOX_DATA_ADDRESS), 0x33334444);

	memory.cartridgeController.reset();
	assert.equal(memory.readMappedU32LE(IO_CART_SELECT), 1);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE), 0x11112222);
	assert.equal(memory.readMappedU32LE(MAILBOX_DATA_ADDRESS), 0);
	assert.equal(memory.readMappedU32LE(MAILBOX_STATUS_ADDRESS), 0);
});

test('mailbox IRQ raises once per cartridge source-latch edge', () => {
	const { memory } = createHarness([
		slot([], CARTRIDGE_BOARD_MAILBOX, 0, true, false),
		slot([], 0, 0, false, false),
	]);

	memory.writeMappedU32LE(MAILBOX_CONTROL_ADDRESS, CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	assert.notEqual(memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT0, 0);
	assert.equal(memory.readMappedU32LE(MAILBOX_STATUS_ADDRESS), CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING);

	memory.writeMappedU32LE(IO_IRQ_ACK, IRQ_CARTRIDGE_SLOT0);
	assert.equal(memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT0, 0);
	memory.writeMappedU32LE(MAILBOX_CONTROL_ADDRESS, CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	assert.equal(
		memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT0,
		0,
		'central IRQ acknowledgement does not clear the cartridge source latch',
	);

	memory.writeMappedU32LE(MAILBOX_IRQ_ACK_ADDRESS, 1);
	assert.equal(memory.readMappedU32LE(MAILBOX_STATUS_ADDRESS), 0);
	memory.writeMappedU32LE(MAILBOX_CONTROL_ADDRESS, CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	assert.notEqual(memory.readMappedU32LE(IO_IRQ_FLAGS) & IRQ_CARTRIDGE_SLOT0, 0);
});

test('cartridge DREQ selectors override CPU selection independently on both DMA sides', () => {
	const board = CARTRIDGE_BOARD_RAM | CARTRIDGE_BOARD_MAILBOX;
	const { memory, dma, scheduler } = createHarness([
		slot([0x01020304, 0x11121314], board, 16, true, false),
		slot([0xa1a2a3a4, 0xb1b2b3b4], board, 16, true, false),
	]);

	memory.writeMappedU32LE(IO_CART_SELECT, 1);
	memory.writeMappedU32LE(
		MAILBOX_CONTROL_ADDRESS,
		CARTRIDGE_MAILBOX_CONTROL_DREQ_READ,
	);
	memory.writeMappedU32LE(IO_CART_SELECT, 0);
	memory.writeMappedU32LE(
		MAILBOX_CONTROL_ADDRESS,
		CARTRIDGE_MAILBOX_CONTROL_DREQ_WRITE,
	);

	const control = DMA_CONTROL_READ_INCREMENT
		| DMA_CONTROL_WRITE_INCREMENT
		| (DMA_REQUEST_CARTRIDGE_SLOT1_READ << DMA_CONTROL_READ_REQUEST_SHIFT)
		| (DMA_REQUEST_CARTRIDGE_SLOT0_WRITE << DMA_CONTROL_WRITE_REQUEST_SHIFT)
		| (1 << DMA_CONTROL_BLOCK_WORDS_SHIFT);
	memory.writeMappedU32LE(IO_DMA0_READ_ADDR, CART_ROM_BASE);
	memory.writeMappedU32LE(IO_DMA0_WRITE_ADDR, CART_RAM_BASE);
	memory.writeMappedU32LE(IO_DMA0_TRANSFER_COUNT, 2);
	memory.writeMappedU32LE(IO_DMA0_CONTROL, control);
	memory.writeMappedU32LE(IO_DMA0_TRIGGER, DMA_TRIGGER_START);
	const deadline = scheduler.nextDeadline();
	assert.equal(deadline, 4);
	scheduler.advanceTo(deadline);
	dma.onService(deadline);

	assert.equal(memory.readMappedU32LE(CART_RAM_BASE), 0xa1a2a3a4);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE + 4), 0xb1b2b3b4);
	memory.writeMappedU32LE(IO_CART_SELECT, 1);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE), 0);
	assert.equal(memory.readMappedU32LE(CART_RAM_BASE + 4), 0);
});

File: tests/cpp/cartridge_controller_test.cpp
#include "spec/bmsx/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/cartridge/contracts.h"
#include "spec/bmsx/cartridge.h"
#include "machine/devices/dma/controller.h"
#include "machine/devices/irq/controller.h"
#include "spec/bmsx/memory_map.h"
#include "machine/memory/memory.h"
#include "spec/bmsx/model.h"
#include "machine/scheduler/device.h"

#include <array>
#include <stdexcept>

namespace {

void require(bool condition, const char* message) {
	if (!condition) {
		throw std::runtime_error(message);
	}
}

struct CartridgeHarness {
	bmsx::Memory memory;
	bmsx::IrqController irq;
	bmsx::ExecutionAddressSpace executionAddressSpace;
	bmsx::CPU cpu;
	bmsx::DeviceScheduler scheduler;
	bmsx::DmaController dma;

	explicit CartridgeHarness(const bmsx::CartridgeSocketMediaPair& slots)
		: memory(bmsx::MemoryInit{ {}, slots }, bmsx::PSX_MACHINE_SPEC.ramBytes)
		, irq(memory)
		, executionAddressSpace(memory)
		, cpu(memory, irq, executionAddressSpace)
		, scheduler(cpu)
		, dma(memory, cpu, irq, scheduler) {
		memory.cartridgeController().connect(memory, irq, dma);
		irq.reset();
		dma.reset();
		memory.cartridgeController().reset();
		dma.setTiming(1, 0, 1, 1, 0, scheduler.nowCycles());
	}
};

void testPhysicalSocketSelectionAndRawLatch() {
	const std::array<bmsx::u8, 4> slot0Rom{{ 0x44u, 0x33u, 0x22u, 0x11u }};
	const std::array<bmsx::u8, 4> slot1Rom{{ 0xddu, 0xccu, 0xbbu, 0xaau }};
	bmsx::CartridgeSocketMediaPair slots{};
	slots[0] = bmsx::CartridgeCardMedia{slot0Rom, 16u, false};
	slots[1] = bmsx::CartridgeCardMedia{slot1Rom, std::nullopt, true};
	CartridgeHarness harness(slots);
	bmsx::Memory& memory = harness.memory;

	require(memory.readMappedU32LE(bmsx::IO_CART_SELECT) == 0u, "reset selects physical socket 0");
	require(memory.readMappedU32LE(bmsx::CART_ROM_BASE) == 0x11223344u, "CPU aperture reads slot 0 after reset");
	require(
		memory.readMappedU32LE(bmsx::IO_CART_STATUS)
			== (bmsx::CARTRIDGE_STATUS_SLOT0_PRESENT
				| bmsx::CARTRIDGE_STATUS_SLOT1_PRESENT),
		"status reports both inserted cartridges"
	);
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 0xa5a50001u);
	require(memory.readMappedU32LE(bmsx::IO_CART_SELECT) == 0xa5a50001u, "selection register retains its raw word");
	require(memory.readMappedU32LE(bmsx::CART_ROM_BASE) == 0xaabbccddu, "CPU aperture reads slot 1 when CS1 is selected");
	require(
		memory.readMappedU32LE(bmsx::IO_CART_STATUS)
			== (bmsx::CARTRIDGE_STATUS_SLOT0_PRESENT
				| bmsx::CARTRIDGE_STATUS_SLOT1_PRESENT
				| bmsx::CARTRIDGE_STATUS_SELECTED_SLOT1),
		"status reflects the selected physical socket"
	);
}

void testAbsentSocketDrivesZeroAndHasNoState() {
	const std::array<bmsx::u8, 4> slot0Rom{{ 0x44u, 0x33u, 0x22u, 0x11u }};
	bmsx::CartridgeSocketMediaPair slots{};
	slots[0] = bmsx::CartridgeCardMedia{slot0Rom, std::nullopt, false};
	CartridgeHarness harness(slots);
	bmsx::Memory& memory = harness.memory;
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	require(
		memory.readMappedU32LE(bmsx::IO_CART_STATUS)
			== (bmsx::CARTRIDGE_STATUS_SLOT0_PRESENT
				| bmsx::CARTRIDGE_STATUS_SELECTED_SLOT1),
		"status retains selection while reporting the second socket absent"
	);
	require(memory.readMappedU32LE(bmsx::CART_ROM_BASE) == 0u, "absent socket drives zero in ROM aperture");
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0u, "absent socket drives zero in RAM aperture");
	require(memory.readMappedU32LE(bmsx::CART_MMIO_BASE) == 0u, "absent socket drives zero in MMIO aperture");
	require(!memory.cartridgeController().captureState().slots[1].has_value(), "absent socket has no synthetic state");
}

void testSocketLocalRamMailboxResetAndRestore() {
	bmsx::CartridgeSocketMediaPair slots{};
	slots[0] = bmsx::CartridgeCardMedia{std::nullopt, 16u, true};
	slots[1] = bmsx::CartridgeCardMedia{std::nullopt, 16u, true};
	CartridgeHarness harness(slots);
	bmsx::Memory& memory = harness.memory;

	require(memory.readMappedU32LE(bmsx::IO_CART_SELECT) == 0u, "cartridge hardware does not interpret executable metadata");
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	memory.writeMappedU32LE(bmsx::CART_RAM_BASE, 0x11112222u);
	memory.writeMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET, 0x33334444u);
	memory.writeMappedU32LE(
		bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET,
		0x80000000u | bmsx::CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER | bmsx::CARTRIDGE_MAILBOX_CONTROL_DREQ_READ
	);
	require(
		memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET)
			== (0x80000000u | bmsx::CARTRIDGE_MAILBOX_CONTROL_DREQ_READ),
		"mailbox control retains representable high bits as a raw word"
	);
	require(
		memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_STATUS_OFFSET)
			== bmsx::CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING,
		"slot 1 mailbox retains its IRQ source latch"
	);
	require((memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT1) != 0u, "slot 1 raises its own IRQ line");

	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 0x2468ace0u);
	memory.writeMappedU32LE(bmsx::CART_RAM_BASE, 0x55556666u);
	memory.writeMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET, 0x77778888u);
	memory.writeMappedU32LE(
		bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET,
		bmsx::CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER | bmsx::CARTRIDGE_MAILBOX_CONTROL_DREQ_WRITE
	);
	require((memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT0) != 0u, "slot 0 raises its own IRQ line");
	const bmsx::CartridgeControllerState saved = memory.cartridgeController().captureState();

	memory.writeMappedU32LE(bmsx::CART_RAM_BASE, 0u);
	memory.writeMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET, 0u);
	memory.writeMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_IRQ_ACK_OFFSET, 1u);
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	memory.cartridgeController().restoreState(saved);

	require(memory.readMappedU32LE(bmsx::IO_CART_SELECT) == 0x2468ace0u, "restore retains the raw selection word");
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0x55556666u, "restore returns slot 0 RAM");
	require(memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET) == 0x77778888u, "restore returns slot 0 mailbox data");
	require(
		memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_STATUS_OFFSET)
			== bmsx::CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING,
		"restore returns slot 0 mailbox IRQ state"
	);
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0x11112222u, "slot 1 RAM is independent");
	require(memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET) == 0x33334444u, "slot 1 mailbox is independent");

	memory.cartridgeController().reset();
	require(memory.readMappedU32LE(bmsx::IO_CART_SELECT) == 0u, "reset returns to physical socket 0");
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0x55556666u, "reset retains cartridge RAM");
	require(memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_DATA_OFFSET) == 0u, "reset clears mailbox data");
	require(memory.readMappedU32LE(bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_STATUS_OFFSET) == 0u, "reset clears mailbox IRQ state");
}

void testMailboxIrqRaisesOncePerSourceLatchEdge() {
	bmsx::CartridgeSocketMediaPair slots{};
	slots[0] = bmsx::CartridgeCardMedia{std::nullopt, std::nullopt, true};
	CartridgeHarness harness(slots);
	bmsx::Memory& memory = harness.memory;
	const bmsx::u32 mailboxControl = bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET;
	const bmsx::u32 mailboxStatus = bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_STATUS_OFFSET;
	const bmsx::u32 mailboxAck = bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_IRQ_ACK_OFFSET;

	memory.writeMappedU32LE(mailboxControl, bmsx::CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	require((memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT0) != 0u, "mailbox edge raises slot IRQ");
	require(
		memory.readMappedU32LE(mailboxStatus) == bmsx::CARTRIDGE_MAILBOX_STATUS_IRQ_PENDING,
		"mailbox source latch records the edge");

	memory.writeMappedU32LE(bmsx::IO_IRQ_ACK, bmsx::IRQ_CARTRIDGE_SLOT0);
	require((memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT0) == 0u, "central IRQ acknowledgement clears the controller flag");
	memory.writeMappedU32LE(mailboxControl, bmsx::CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	require(
		(memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT0) == 0u,
		"central IRQ acknowledgement does not clear the cartridge source latch");

	memory.writeMappedU32LE(mailboxAck, 1u);
	require(memory.readMappedU32LE(mailboxStatus) == 0u, "local acknowledgement clears the mailbox source latch");
	memory.writeMappedU32LE(mailboxControl, bmsx::CARTRIDGE_MAILBOX_CONTROL_IRQ_TRIGGER);
	require((memory.readMappedU32LE(bmsx::IO_IRQ_FLAGS) & bmsx::IRQ_CARTRIDGE_SLOT0) != 0u, "a new mailbox edge raises the IRQ again");
}

void testDmaRequestSelectorsOverrideBothChipSelects() {
	const std::array<bmsx::u8, 8> slot0Rom{{
		0x04u, 0x03u, 0x02u, 0x01u,
		0x14u, 0x13u, 0x12u, 0x11u,
	}};
	const std::array<bmsx::u8, 8> slot1Rom{{
		0xa4u, 0xa3u, 0xa2u, 0xa1u,
		0xb4u, 0xb3u, 0xb2u, 0xb1u,
	}};
	bmsx::CartridgeSocketMediaPair slots{};
	slots[0] = bmsx::CartridgeCardMedia{slot0Rom, 16u, true};
	slots[1] = bmsx::CartridgeCardMedia{slot1Rom, 16u, true};
	CartridgeHarness harness(slots);
	bmsx::Memory& memory = harness.memory;

	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	memory.writeMappedU32LE(
		bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET,
		bmsx::CARTRIDGE_MAILBOX_CONTROL_DREQ_READ
	);
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 0u);
	memory.writeMappedU32LE(
		bmsx::CART_MMIO_BASE + bmsx::CARTRIDGE_MAILBOX_CONTROL_OFFSET,
		bmsx::CARTRIDGE_MAILBOX_CONTROL_DREQ_WRITE
	);

	const bmsx::u32 control = bmsx::DMA_CONTROL_READ_INCREMENT
		| bmsx::DMA_CONTROL_WRITE_INCREMENT
		| (bmsx::DMA_REQUEST_CARTRIDGE_SLOT1_READ << bmsx::DMA_CONTROL_READ_REQUEST_SHIFT)
		| (bmsx::DMA_REQUEST_CARTRIDGE_SLOT0_WRITE << bmsx::DMA_CONTROL_WRITE_REQUEST_SHIFT)
		| (1u << bmsx::DMA_CONTROL_BLOCK_WORDS_SHIFT);
	memory.writeMappedU32LE(bmsx::IO_DMA0_READ_ADDR, bmsx::CART_ROM_BASE);
	memory.writeMappedU32LE(bmsx::IO_DMA0_WRITE_ADDR, bmsx::CART_RAM_BASE);
	memory.writeMappedU32LE(bmsx::IO_DMA0_TRANSFER_COUNT, 2u);
	memory.writeMappedU32LE(bmsx::IO_DMA0_CONTROL, control);
	memory.writeMappedU32LE(bmsx::IO_DMA0_TRIGGER, bmsx::DMA_TRIGGER_START);
	const bmsx::i64 deadline = harness.scheduler.nextDeadline();
	require(deadline == 4, "cartridge read and write serialize on the shared external bus");
	harness.scheduler.advanceTo(deadline);
	harness.dma.onService(deadline);

	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0xa1a2a3a4u, "DMA read selector drives slot 1 CS");
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE + 4u) == 0xb1b2b3b4u, "DMA burst retains slot 1 CS");
	memory.writeMappedU32LE(bmsx::IO_CART_SELECT, 1u);
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE) == 0u, "DMA write selector drives slot 0 CS independently");
	require(memory.readMappedU32LE(bmsx::CART_RAM_BASE + 4u) == 0u, "slot 1 RAM is not the DMA destination");
}

} // namespace

int main() {
	testPhysicalSocketSelectionAndRawLatch();
	testAbsentSocketDrivesZeroAndHasNoState();
	testSocketLocalRamMailboxResetAndRestore();
	testMailboxIrqRaisesOncePerSourceLatchEdge();
	testDmaRequestSelectorsOverrideBothChipSelects();
	return 0;
}
